homeassistant/components/mikrotik/__init__.py

```diff
@@ -1 +1,259 @@
+import ssl
+
+import voluptuous as vol
+import librouteros
+from librouteros.login import login_plain, login_token
+
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_METHOD,
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+from homeassistant.components.device_tracker import DOMAIN as DEVICE_TRACKER
+from .const import (
+    DOMAIN,
+    MTK_LOGIN_PLAIN,
+    MTK_LOGIN_TOKEN,
+    DEFAULT_ENCODING,
+    IDENTITY,
+    CONF_TRACK_DEVICES,
+    CONF_ENCODING,
+    CONF_ARP_PING,
+    CONF_LOGIN_METHOD,
+    INFO,
+    MIKROTIK_SERVICES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+MTK_DEFAULT_API_PORT = "8728"
+MTK_DEFAULT_API_SSL_PORT = "8729"
+
+MIKROTIK_SCHEMA = vol.All(
+    vol.Schema(
+        {
+            vol.Required(CONF_HOST): cv.string,
+            vol.Required(CONF_USERNAME): cv.string,
+            vol.Required(CONF_PASSWORD): cv.string,
+            vol.Optional(CONF_METHOD): cv.string,
+            vol.Optional(CONF_LOGIN_METHOD): vol.Any(MTK_LOGIN_PLAIN, MTK_LOGIN_TOKEN),
+            vol.Optional(CONF_PORT): cv.port,
+            vol.Optional(CONF_SSL, default=False): cv.boolean,
+            vol.Optional(CONF_ENCODING, default=DEFAULT_ENCODING): cv.string,
+            vol.Optional(CONF_TRACK_DEVICES, default=True): cv.boolean,
+            vol.Optional(CONF_ARP_PING, default=False): cv.boolean,
+        }
+    )
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [MIKROTIK_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Mikrotik component."""
+    hass.data[DOMAIN] = {}
+
+    for device in config[DOMAIN]:
+        host = device[CONF_HOST]
+        use_ssl = device.get(CONF_SSL)
+        user = device.get(CONF_USERNAME)
+        password = device.get(CONF_PASSWORD, "")
+        login = device.get(CONF_LOGIN_METHOD)
+        encoding = device.get(CONF_ENCODING)
+        method = device.get(CONF_METHOD)
+        arp_ping = device.get(CONF_ARP_PING)
+        track_devices = device.get(CONF_TRACK_DEVICES)
+
+        if CONF_PORT in device:
+            port = device.get(CONF_PORT)
+        else:
+            if use_ssl:
+                port = MTK_DEFAULT_API_SSL_PORT
+            else:
+                port = MTK_DEFAULT_API_PORT
+
+        if login == MTK_LOGIN_PLAIN:
+            login_method = (login_plain,)
+        elif login == MTK_LOGIN_TOKEN:
+            login_method = (login_token,)
+        else:
+            login_method = (login_plain, login_token)
+
+        try:
+            api = MikrotikClient(
+                host,
+                use_ssl,
+                port,
+                user,
+                password,
+                login_method,
+                encoding,
+                arp_ping,
+                track_devices,
+            )
+            api.connect_to_device()
```

Comment:
This call isn't async safe. The librouteros library doesn't support asyncio.

We should switch to use the home assistant sync api instead of the async api. Ie ue `setup` instead of `async_setup` etc.

Reply:
Okay, I will update the code to remove the async calls. Does that mean the device tracker also needs async removed?

Reply:
It's simpler to replace all async calls with sync, yes. Technically it depends on what is done in the calls. Normal I/O or other blocking behavior is forbidden in the event loop.